### DISCUSSION TITLE
Question about determining whether face neighbors are local

### DISCUSSION THREAD

**Discussion by jthano** [chi-tech/chi-tech#95] [Q&A] [answered]
The CellFace class has
```c++
bool chi_mesh::CellFace::
  IsNeighborLocal(chi_mesh::MeshContinuum& grid) const
```
Now `CellLBSView` also caches whether a face has a local neighbor in `std::vector<bool> face_local = {};` Looking at `InitializeParrays()`, I think the data in CellLBSVeiw is just a cache of the data in CellFace. But I wanted to make sure of this since the CellLBSVeiw version is used in the sweepchunk object.

Is `CellLBSVeiw::face_local` always meant to contain the same information as is in `CellFace::IsNeighborLocal`?

**Comment by Naktakala** [comment, marked as answer]
Very good question!

`CellFace::IsNeighborLocal` is a method call that needs a reference to the grid. This then checks whether the face has a neighbor and if it does it accesses the grid's global-cell-handler aka `grid.cells`. The latter is not a mere array but an accessor to two maps (one for local cells and one for non-local/ghost cells), which will be checked to see if the request is valid and return either the local-neighbor or ghost-neighbor which in turn can be used to check `adj_cell.partition_id == chi_mpi.location_id`. 

One of the first design-goals I had in mind was to keep everything as simple and lightweight as possible, this meant I didn't want to load up cells and faces with a ton of information that needed to be initialized in a very particular manner (especially if only a handful of methods actually required it). That is why that information is not stored. 

Another design-goal, is speed. Sweeps need to check face locality every time a sweep chunk executes, making the additional checks from `CellFace::IsNeighborLocal` a very time consuming thing and something that can be done once-off.

Another purpose of the `CellLBSView` is DOF mapping. The block address of a cell has a bunch of multiplications part of which can be stored `dof_phi_map_start`. 

So in conclusion. CellLBSView is just a container for some caching yes!
